### ISSUE TITLE
`drop` on non-`Drop` types is useful defensive programming

### ISSUE THREAD

**Issue by wchargin** [rust-lang/rust-clippy#11650]
### Summary

If `v` is a value that is not `Copy`, then `drop(v)` ensures that `v` is
not used after that point. A later change to the code that accidentally
reuses that variable will lead to a compiler error, preventing bugs
especially when there are multiple variables of the same type in scope.

(Morally, this is similar to shadowing, and indeed a similar effect can
be obtained by replacing `drop(v)` with `let () = v`.)

However, if this technique is used when `v` does not have a nontrivial
destructor, then Clippy complains with the `drop_non_drop` lint. This is
bad, because accepting the lint's suggestion makes the code less robust
to future changes.

### Lint Name

drop_non_drop

### Reproducer

I tried this code:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct State {
    x: u64,
}

pub enum Op {
    Update(u64),
    Read,
}

// This function takes a complex set of parameters (here `Op`). It reads
// from `state` and, depending on its parameters, may or may not also mutate it.
pub fn operate(op: Op, state: &mut State) -> u64 {
    match op {
        Op::Update(y) => state.x += y,
        Op::Read => (),
    };
    state.x
}

pub fn do_multiple_operations(state: &mut State) -> u64 {
    let mut temp_state = state.clone();
    let a = operate(Op::Read, &mut temp_state);
    // Double-check that we didn't mutate the state, then discard the temp state
    // so that we don't accidentally use it instead of the real one.
    assert_eq!(&temp_state, state);
    drop(temp_state);

    // If we accidentally use `&mut temp_state` instead of `state` on the next
    // line, we get an error, because we dropped it above (good).
    let b = operate(Op::Update(77), state);
    a + b
}
```

I saw this happen:

```
warning: call to `std::mem::drop` with a value that does not implement `Drop`. Dropping such a type only extends its contained lifetimes
  --> src/lib.rs:27:5
   |
27 |     drop(temp_state);
   |     ^^^^^^^^^^^^^^^^
   |
note: argument has type `State`
  --> src/lib.rs:27:10
   |
27 |     drop(temp_state);
   |          ^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#drop_non_drop
   = note: `#[warn(clippy::drop_non_drop)]` on by default

warning: `playground` (lib) generated 1 warning
```

I expected to see this happen:

No lint warning, because the code is useful and working as intended.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```

**Comment by Jarcho** [comment]
One reason for the lint is to catch the case where `drop(x)` was used instead of `drop(y)`. If you do want the behaviour you can use `let _ = x` (not `let () = x`, I assume a typo on your part). Basically the lint is enforcing the idea that explicitly dropping a value should actually do something.

**Comment by Centri3** [comment]
Isn't `let _ = x` a no-op? But I think they meant `let x = ()` since this shadows it (though clippy disagrees with this currently)

**Comment by Jarcho** [comment]
So it is. The first point about mixing up the dropped value still stands, but it does leave no easy way do this without getting linted.

**Comment by wchargin** [comment]
Er, yes, thanks, I meant `let x = ()`.

It is indeed unfortunate that Clippy dislikes this, too.

**Comment by asquared31415** [comment]
Adding additional scopes can be used to prevent names from being available in places that you don't want.
```rust
let a = {
    let mut temp_state = state.clone();
    modify_state(&mut temp_state);
    temp_state
};
```

**Comment by Jarcho** [comment]
Another option would be to do something like `let _dont_use = temp;`. Combined with `used_underscore_bindings` this will also warn when the value is used.

`drop_non_drop` isn't likely to change. It's in the `suspicious` category because it catches bugs.

**Comment by wchargin** [comment]
At the very least, it is strictly untrue that "Dropping such a type only extends its contained lifetimes". Dropping such a non-`Copy` type _has_ another effect—making it unusable after that point.

**Comment by RmbRT** [comment]
What if you want to drop something exactly to free up the references it holds in its fields? As far as I know, drop() exactly exists for the express purpose of dropping values in a semantically expressive way, instead of writing `let _ = x;` (or `{let _me_please_drop = x;}` or whatever). Sometimes, the borrow checker demands I drop some object, and sometimes I have to drop an outer object in an inner scope, to make references it holds available in outer scopes again. Clippy has only produced false positives or trivial nagging for my code so far. Please fix it, instead of going against best practices and idioms of the base language. Especially since I get nagged in code review for not doing things the idiomatic way, because Rust is so big on forcing everyone into being idiomatic.

**Comment by Jarcho** [comment]
Lifetimes held by a type are only held until the value's last use. For types with drop glue that will be when it's dropped, but for all other types adding an explicit drop doesn't shorten any lifetimes.

**Comment by WorldSEnder** [comment]
Explicit drop can still be used to shorten an associated lifetime. It doesn't shorten the lifetime in the type of the dropped variable, but it can shorten others:

```rust
let guard = ();
let associated = &guard; // &'1 ()
// ... some code
drop(guard); // With this drop, we guarantee that the lifetime of the borrow ('1) ends before this point.
```

This can be especially useful with `unsafe` code, where a reference with the inferred lifetime might leak past the point of the `drop` which the explicit end can help catch.
